### PR TITLE
🧱 Basic architecture

### LINKED ISSUE #1: 🧱 Create base interface architecture
As viewed in meeting the project will lay down on some interfaces that should cover all basic needs and prevents future breaking changes.

# Game Interface
The Game Interface will represent each future game. We defined the game as something that can, start, run and have a score, so the interface should have at least those functions.
The Game Interface may also need a function to link a game to an arcade instance.

# Arcade Interface
The arcade, even if handled in the group scope, has to have common functions.
The arcade will link games to graphic drivers in a “one way” connection from the game to the driver (There is an exception for bindings and event, but that's about it), so the arcade interface will need at least the same functions as an IDriver

# Driver Interface
A graphic driver interface (aka IDriver) is what will *interface* with each graphic library. It should accept three main things:
1. A display function that accepts a displayable object.
2. A function that flip a frame. (Flipping a frame shall be considered as the process of freezing the buffered frame, then displaying it on the screen and at last create a new **clean** buffered frame for future draws)
3. A function that binds event. Each graphic library will handle this in their own way, but a game is interactive at the core. Each game shall be able to bind callbacks to available keys that will be called when the graphic library receives the event.

# Entities, Sprites & Primitives
Entities and Primitives are both displayable object, so we have to create a parent interface for those both that has some getter for the position and the size.

## Primitives
For simpler future development, we define primitives as shapes that have a color. Moreover, many graphic library gives function to draw specific shapes, such as circle, square and/or triangles so we shall create such interfaces giving radius or sides length for example.

## Entities
Entities are a bit more complex whilst staying a displayable object. An entity has a sprite attached to it, but later we could add more information like material, opacity etc...

## Sprites
A Sprite has a function to get the image for the actual frame, but also a char and a color as a fallback for a library that can't draw images or if the image is not available.

# Images
Since each library defined its own image type, we shall create a base level representation of what an image really is. This interface shall have getters for size and path at least.

### PR DESCRIPTION
**Pull request by axel-eck**
I made a basic interface architecture following what was laid down on #1.

### CODE REVIEW

File: utils/IPicture.hpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+** No file there , just an epitech header example .
+** You can even have multiple lines if you want !
+*/
+
+#pragma once
+#include <string>
+
+class IPicture {
+public:
+    string getPath() = 0;
```
Comment (Antoinegtir) [inline utils/IPicture.hpp]: make getPath function virtual and allow it to return std::string

File: utils/IPicture.hpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/*
+** EPITECH PROJECT, 2024
+** ArcadeShared
+** File description:
+** No file there , just an epitech header example .
+** You can even have multiple lines if you want !
+*/
+
+#pragma once
+#include <string>
```
Comment (Antoinegtir) [inline utils/IPicture.hpp:10]: change include string to include <iostream>

File: utils/IPicture.hpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+class IPicture {
+public:
+    string getPath() = 0;
+    void setPath(string path) = 0;
```
Comment (Antoinegtir) [inline utils/IPicture.hpp]: make setPath virtual and pass as parameter std::string

File: displayable/entities/ISprite.hpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+** You can even have multiple lines if you want !
+*/
+
+#pragma once
+#include <memory>
+#include "../../utils/IPicture.hpp"
+
+class ISprite {
+public:
+    std::shared_ptr<IPicture> getPicture() = 0;
```
Comment (Antoinegtir) [inline displayable/entities/ISprite.hpp]: add virtual key word for getPicture

File: displayable/entities/ISprite.hpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    void setPicture(std::unique_ptr<IPicture> picture) = 0;
```
Comment (Antoinegtir) [inline displayable/entities/ISprite.hpp]: add virtual key word for setPicture

File: utils/IPicture.hpp (hunk 2 above)
Comment (axel-eck) [reply utils/IPicture.hpp:10]: It's not necessary to include the whole "iostream" header since it itself includes "string" and that's only what we need.

**Review by Neiluge — APPROVED** [review]

**Review by aimericdrk — APPROVED** [review]

**Review by Antoinegtir — APPROVED** [review]

**Review by Marius-P1 — APPROVED** [review]
Good for me :ok:

**Pull request merged**